// angular2-bootstrap-confirm/dist/esm/src/confirmOptions.provider.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { Injectable } from '@angular/core';
/**
 * Use this for configuring the confirmation popover defaults
 *
 * Use it like so:
 * ```
 * const options: ConfirmOptions = new ConfirmOptions();
 * // focus the confirm button by default. See below for an explanation of all options.
 * options.focusButton = 'confirm';
 *
 * providers: [{provide: ConfirmOptions, useValue: options}]
 * ```
 */
export var ConfirmOptions = (function () {
    function ConfirmOptions() {
        /**
         * The popover confirmation button text
         */
        this.confirmText = 'Confirm';
        /**
         * The popover cancel button text
         */
        this.cancelText = 'Cancel';
        /**
         * The popover confirm button type e.g. `success`, `danger` etc
         */
        this.confirmButtonType = 'success';
        /**
         * The popover cancel button type  e.g. `success`, `danger` etc
         */
        this.cancelButtonType = 'default';
        /**
         * The popover placement. Can be `top`, `bottom`, `left`, `right`
         */
        this.placement = 'top';
        /**
         * Whether to hide the confirmation button
         */
        this.hideConfirmButton = false;
        /**
         * Whether to hide the cancel button
         */
        this.hideCancelButton = false;
        /**
         * A custom CSS class to be added to the popover
         */
        this.popoverClass = '';
        /**
         * Whether to append the popover to the document body
         */
        this.appendToBody = false;
    }
    ConfirmOptions.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ConfirmOptions.ctorParameters = [];
    return ConfirmOptions;
}());
/**
 * @private
 */
export var PopoverConfirmOptions = (function (_super) {
    __extends(PopoverConfirmOptions, _super);
    function PopoverConfirmOptions() {
        _super.apply(this, arguments);
    }
    PopoverConfirmOptions.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    PopoverConfirmOptions.ctorParameters = [];
    return PopoverConfirmOptions;
}(ConfirmOptions));
